synapse/lib/stormlib/imap.py

```diff
@@ -0,0 +1,376 @@
+
+                Mark a set of messages of messages as deleted but do not expunge::
+
+                    ($ok, $valu) = $server.markDeleted("1,4,6", expunge=$lib.false)
+            ''',
+            'type': {
+                'type': 'function', '_funcname': 'markDeleted',
+                'args': (
+                    {'type': 'str', 'name': 'uid_set',
+                     'desc': 'The UID message set to apply the flag to.'},
+                    {'type': 'boolean', 'name': 'expunge', 'default': True,
+                     'desc': 'Execute the expunge command on the mailbox after setting the flag.'},
+                ),
+                'returns': {
+                    'type': 'list',
+                    'desc': 'An ($ok, $valu) tuple.'
+                },
+            },
+        },
+    )
+    _storm_typename = 'storm:imap:server'
+
+    def __init__(self, runt, imap_cli, path=None):
+        s_stormtypes.StormType.__init__(self, path=path)
+        self.runt = runt
+        self.imap_cli = imap_cli
+        self.locls.update(self.getObjLocals())
+
+    def getObjLocals(self):
+        return {
+            'list': self.list,
+            'fetch': self.fetch,
+            'login': self.login,
+            'search': self.search,
+            'select': self.select,
+            'markSeen': self.markSeen,
+            'markDeleted': self.markDeleted,
+        }
+
+    async def login(self, user, passwd):
+        user = await s_stormtypes.tostr(user)
+        passwd = await s_stormtypes.tostr(passwd)
+
+        coro = self.imap_cli.login(user, passwd)
+        await run_imap_coro(coro)
+
+        return True, None
+
+    async def list(self, reference_name='""', pattern='*'):
+        pattern = await s_stormtypes.tostr(pattern)
+        reference_name = await s_stormtypes.tostr(reference_name)
+
+        coro = self.imap_cli.list(reference_name, pattern)
+        data = await run_imap_coro(coro)
+
+        names = []
+        for item in data:
+            if item == b'Success':
+                break
+            names.append(item.split(b' ')[-1].decode().strip('"'))
+
+        return True, names
+
+    async def select(self, mailbox='INBOX'):
+        mailbox = await s_stormtypes.tostr(mailbox)
+
+        coro = self.imap_cli.select(mailbox=mailbox)
+        await run_imap_coro(coro)
+
+        return True, None
+
+    async def search(self, *args):
+        args = [await s_stormtypes.tostr(arg) for arg in args]
+
+        coro = self.imap_cli.uid_search(*args)
+        data = await run_imap_coro(coro)
+
+        uids = data[0].decode().split(' ') if data[0] else []
+        return True, uids
+
+    async def fetch(self, uid):
+        '''
+        IMAP fetch accepts a message set (e.g. "1", "1:*", "1,2,3"),
+        however this method forces fetching a single uid
+        to prevent retrieving a very large blob of data.
+        '''
+
+        self.runt.confirm(('storm', 'lib', 'imap', 'fetch'))
```

Comment:
why is this the only method with a permission gate?

Reply:
To gate storing to the axon.

Reply:
Lets add a perm check to `connect()` to gate the interaction as a whole and drop this one.

Reply:
updated